runtime/drive_handler.go

```diff
@@ -0,0 +1,192 @@
+	//
+	// This should occur less and less, unless the amount is constantly
+	// increasing.
+	h.Lock.RUnlock()
+	h.Lock.Lock()
+	defer h.Lock.Unlock()
+	if h.DriveAmount >= amount {
+		return h.Paths, nil
+	}
+
+	for ; h.DriveAmount < amount; h.DriveAmount++ {
+		driveID := fmt.Sprintf("stub%d", h.DriveAmount)
+		path := filepath.Join(h.RootPath, driveID)
+
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_RDWR, 0600)
+		if err != nil {
+			return nil, err
+		}
+
+		if _, err = f.WriteString(fmt.Sprintf("%s%s", internal.MagicStubBytes, driveID)); err != nil {
+			f.Close()
+			return nil, err
+		}
+
+		// Firecracker will not show any drives smaller than 512 bytes. If the drive
+		// size is not a multiple of 512, then those bytes will not be visible to
+		// firecracker.
+		if err := os.Truncate(path, minFCDriveImageBytes); err != nil {
+			f.Close()
+			return nil, err
+		}
+
+		h.Paths = append(h.Paths, path)
+		f.Close()
+	}
+
+	return h.Paths, nil
+}
+
+// SetDrives will set the given drives based off the VM id. The index
+// represents where the stub drive index starts.
+func (h *stubDriveHandler) SetDrives(id string, index int, d []models.Drive) {
+	h.drives.Store(id, driveManager{
+		stubDriveIndex: index,
+		drives:         d,
+	})
+}
+
+// GetDrives will return a series of drives related to the given key, which in
+// this case is the VM ID.
+func (h *stubDriveHandler) GetDrives(id string) ([]models.Drive, bool) {
+	drivesIface, ok := h.drives.Load(id)
+	if !ok {
+		return nil, false
+	}
+
+	return drivesIface.(driveManager).drives, true
+}
+
+// InDriveSet will attempt to load the drives associated with the given VM ID.
+// If no VM ID is associated with anything, then false will be returned.  Also,
+// if the drive path is the same as any of the paths in the drives, then true
+// will be returned.
+func (h *stubDriveHandler) InDriveSet(id, path string) bool {
+	drives, ok := h.GetDrives(id)
+	if !ok {
+		return false
+	}
+
+	for _, d := range drives {
+		if firecracker.StringValue(d.PathOnHost) == path {
+			return true
+		}
+	}
+
+	return false
+}
+
+// PatchStubDrive will patch a given drive with the next available stub drive.
+func (h *stubDriveHandler) PatchStubDrive(ctx context.Context, client firecracker.MachineIface, id, pathOnHost string) (*string, error) {
+	drivesIface, ok := h.drives.Load(id)
+	if !ok {
+		return nil, ErrVMNotFound
+	}
+
+	drives := drivesIface.(driveManager)
+	if drives.stubDriveIndex >= len(drives.drives) {
+		return nil, ErrStubDrivesExhausted
+	}
+
+	d := drives.drives[drives.stubDriveIndex]
+	d.PathOnHost = &pathOnHost
+
+	if d.DriveID == nil {
+		// this should never happen, but we want to ensure that we never nil
+		// dereference
+		return nil, ErrDriveIDNil
+	}
+
+	drives.drives[drives.stubDriveIndex] = d
+
+	err := client.UpdateGuestDrive(ctx, firecracker.StringValue(d.DriveID), pathOnHost)
+	if err != nil {
+		return nil, err
+	}
+
+	drives.stubDriveIndex++
```

Comment:
This looks like a race to me.  If multiple calls to `PatchStubDrive` happen at the same time for the same VM, you can have two different goroutines updating stubDriveIndex at the same time and one of them will win.

Reply:
Ah, keep forgetting that these are tied to an API and can have races ><